projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java

```diff
@@ -0,0 +1,320 @@
+import com.oracle.truffle.llvm.parser.base.model.types.ArrayType;
+import com.oracle.truffle.llvm.parser.base.model.types.FloatingPointType;
+import com.oracle.truffle.llvm.parser.base.model.types.FunctionType;
+import com.oracle.truffle.llvm.parser.base.model.types.IntegerType;
+import com.oracle.truffle.llvm.parser.base.model.types.MetaType;
+import com.oracle.truffle.llvm.parser.base.model.types.PointerType;
+import com.oracle.truffle.llvm.parser.base.model.types.StructureType;
+import com.oracle.truffle.llvm.parser.base.model.types.Type;
+import com.oracle.truffle.llvm.parser.base.model.types.VectorType;
+
+public final class LLVMToBitcodeAdapter {
+
+    private LLVMToBitcodeAdapter() {
+    }
+
+    public static Type[] resolveTypes(ResolvedType[] types) {
+        if (types == null) {
+            return null;
+        }
+
+        Type[] resolvedTypes = new Type[types.length];
+        for (int i = 0; i < resolvedTypes.length; i++) {
+            resolvedTypes[i] = resolveType(types[i]);
+        }
+        return resolvedTypes;
+    }
+
+    public static Type resolveType(ResolvedType type) {
+        if (type == null) {
+            return null;
+        }
+
+        if (type instanceof ResolvedNamedType) {
+            return resolveType((ResolvedNamedType) type);
+        } else if (type.isFunction()) {
+            return resolveType((ResolvedFunctionType) type);
+        } else if (type.isFloating()) {
+            return resolveType((ResolvedFloatingType) type);
+        } else if (type.isInteger()) {
+            return resolveType((ResolvedIntegerType) type);
+        } else if (type.isMetadata()) {
+            return resolveType((ResolvedMetadataType) type);
+        } else if (type.isPointer()) {
+            return resolveType((ResolvedPointerType) type);
+        } else if (type.isStruct()) {
+            return resolveType((ResolvedStructType) type);
+        } else if (type.isVararg()) {
+            throw new AssertionError("varargs are only expected inside functions");
+        } else if (type.isVector()) {
+            return resolveType((ResolvedVectorType) type);
+        } else if (type.isVoid()) {
+            return resolveType((ResolvedVoidType) type);
+        } else if (type.isUnknown()) {
+            return resolveType((ResolvedUnknownType) type);
+        } else if (type instanceof ResolvedArrayType) {
+            return resolveType((ResolvedArrayType) type);
+        } else if (type instanceof ResolvedOpaqueType) {
+            return resolveType((ResolvedOpaqueType) type);
+        }
+
+        throw new AssertionError("Unknown type: " + type + " - " + type.getClass().getTypeName());
+    }
+
+    private static Map<ResolvedNamedType, Type> namedTypes = new HashMap<>();
+
+    public static Type resolveType(ResolvedNamedType type) {
+        if (!namedTypes.containsKey(type)) {
+            namedTypes.put(type, MetaType.UNKNOWN); // TODO: resolve cycles
```

Comment:
Nevermind, I see that the resolveType call in the second put might access the value stored in the first.